src/core/detail/btree_internal.h

```diff
@@ -0,0 +1,458 @@
+//   few bytes with b-tree. The trick is using significantly large nodes (256 bytes) so that
+//   their overhead is compared to the items they store.
+//   Why not use absl::btree_set? We must have Rank tree functionality that absl does not supply.
+//   Hacking into absl is not a simple task, implementing our own tree is easier.
+// Below some design decisions:
+// 1. We use predefined node size of 256 bytes and derive number of items in each node from it.
+//    Inner nodes have less items than leaf nodes because they also need to store child pointers.
+// 2. BTreeNode does not predeclare fields besides the 8 bytes metadata - everything else is
+//    calculated at run-time and has dedicated accessors (similarly to absl). This allows
+//    dense and efficient representation of tree nodes.
+// 3. We assume that we store small items (8, 16 bytes) which will have a large branching
+//    factor (248/16), meaning the tree will stay shallow even for sizes reaching billion nodes.
+// 4. We do not store parent pointer like in absl tree. Instead we use BTreePath to store
+//    hierarchy of parent nodes. That should reduce our overhead even further by few bits per item.
+// 5. We assume we store POD types - this greatly reduces the complexity of the generics
+//    in the code.
+// 6. We support pmr memory resource. This allows us to use pluggable heaps.
+//
+// TODO: (all the ideas taken from absl implementation)
+//       1. to introduce slices when removing items from the tree (avoid shifts).
+//       2. to avoid merging/rebalancing when removing max/min items from the tree.
+//       3. Small tree optimization: when the tree is small with a single root node, we can
+//          allocate less then 256 bytes (special case) to avoid relative blowups in memory for
+//          small trees.
+
+template <typename Policy> class BTreeNode {
+  using _Key = typename Policy::Key_t;
+
+  friend class ::dfly::BTree<_Key, Policy>;
+
+  BTreeNode(bool leaf) : val_(0) {
+    leaf_ = leaf;
+  }
+
+  static constexpr uint16_t kKeySize = sizeof(_Key);
+  static constexpr uint16_t kTargetNodeSize = 256;
+  static constexpr uint16_t kKeyOffset = sizeof(uint64_t);  // 8 bytes for metadata
+
+ public:
+  static constexpr uint16_t kMaxLeafKeys = (kTargetNodeSize - kKeyOffset) / kKeySize;
+  static constexpr uint16_t kMinLeafKeys = kMaxLeafKeys / 2;
+
+  // internal node:
+  // x slots, (x+1) children: x * kKeySize + (x+1) * sizeof(BTreeNode*) = x * (kKeySize + 8) + 8
+  // x = (kTargetNodeSize - 8 - kKeyOffset) / (kKeySize + 8)
+
+  static constexpr uint16_t kMaxInnerKeys =
+      (kTargetNodeSize - sizeof(void*) - kKeyOffset) / (kKeySize + sizeof(void*));
+  static constexpr uint16_t kMinInnerKeys = kMaxInnerKeys / 2;
+
+  using Key_t = _Key;
+
+  void InitSingle(Key_t key) {
+    SetKey(0, key);
+    num_items_ = 1;
+  }
+
+  Key_t Key(unsigned index) const {
+    Key_t res;
+    memcpy(&res, KeyPtr(index), kKeySize);
```

Comment:
You could provide (instead or in addition) a `KeyRef()` which returns a reference, and then here you'd do a simple assign.

Reply:
Agree. If the offset will stay 8 bytes in the future, the keys will be aligned correctly so they can be read and written in-place

Reply:
it's very fragile and as @dranikpg mentioned, the alignment issues can easily make it UB.  I got bit by it once: https://www.romange.com/2017/09/26/how-to-serialize-integers-into-memory/

never again 😄 in c++20 they introduce https://en.cppreference.com/w/cpp/numeric/bit_cast  but as you can see it's also not by reference.

Reply:
But if you keep track of alignment it greatly simplifies the deal... Its not complicated, really. We sometimes already do those kind of conversion in serialization code

Sad to see this C feature is not part of C++, would fit in here very well
https://en.wikipedia.org/wiki/Flexible_array_member

Reply:
actually, we can use zero-length array. But it works only for a single array, say "keys". But then children will still be handled like we do it now. I prefer not to use half-measure.

Reply:
I do not understand what's the simplification. Is it to use `KeyT& Key()` instead of using `SetKey` ? 
I do not see it as a benefit but merely a personal style.